tests/combine_reports.py

```diff
@@ -0,0 +1,243 @@
+            "time": float(testsuite.attrib["time"]),
+            "timestamp": testsuite.attrib["timestamp"],
+            "hostname": testsuite.attrib["hostname"],
+            "protocol": protocol,
+            "slave": slave,
+            "testcases": {}
+        } 
+        testcases = testsuite.findall(XML_TESTCASE)
+        
+        for testcase in testcases:
+            classname = testcase.attrib["classname"]
+            name = testcase.attrib["name"]
+            if classname not in test_reports[input_path]["testcases"]:
+                test_reports[input_path]["testcases"][classname] = {}
+            
+            failure = testcase.find("failure")
+            skipped = testcase.find("skipped")
+            error = testcase.find("error")
+            if failure is not None:
+                test_reports[input_path]["testcases"][classname][name] = {
+                    "result": "FAILED",
+                    "message": failure.attrib["message"],
+                    "output": failure.text
+                }
+            elif skipped is not None:
+                test_reports[input_path]["testcases"][classname][name] = {
+                    "result": "SKIPPED",
+                    "message": skipped.attrib["message"],
+                    "output": skipped.text
+                }
+            elif error is not None:
+                test_reports[input_path]["testcases"][classname][name] = {
+                    "result": "ERROR",
+                    "message": error.attrib["message"],
+                    "output": error.text
+                }
+            else:
+                test_reports[input_path]["testcases"][classname][name] = {
+                    "result": "PASSED",
+                    "message": "",
+                    "output": ""
+                }
+            test_reports[input_path]["testcases"][classname][name]["time"] = float(testcase.attrib["time"])
+        
+    return test_reports
+
+
+def combine_reports(test_reports):
+    combined_report = {
+        "time": 0,
+        "errors": 0,
+        "failures": 0,
+        "passed": 0,
+        "skipped": 0,
+        "tests": 0,
+        "testcases": {}
+    }
+    for report_path, report_dict in test_reports.items():
+        protocol = report_dict["protocol"]
+        slave = report_dict["slave"]
+        if "timestamp" not in combined_report:
+            combined_report["timestamp"] = report_dict["timestamp"]
+            combined_report["hostname"] = report_dict["hostname"]
+        combined_report["time"] += report_dict["time"]
+        combined_report["tests"] = report_dict["tests"]
+
+        for classname in report_dict["testcases"].keys():
+            if classname not in combined_report["testcases"]:
+                combined_report["testcases"][classname] = {}
+            for name in report_dict["testcases"][classname].keys():
+                if name not in combined_report["testcases"][classname]:
```

Comment:
It's hard to understand what means these if and else, can you create some functions to encapsulate some of this logic? the function name will help to understand what it does

Reply:
Ok, I just created a function to encapsulate this logic.